### ISSUE TITLE
Workspace diagnostics

### ISSUE THREAD

**Issue by SchnozzleCat** [seblj/roslyn.nvim#32]
Is there a way to get workspace diagnostics to work? Not having to repeatedly run ``dotnet build`` to find errors while refactoring would be nice.

**Comment by SchnozzleCat** [comment]
Looking into this a bit more, it seems the ``roslyn`` server already supports this, but setting the following:

```
["csharp|background_analysis"] = {
  dotnet_compiler_diagnostics_scope = "fullSolution"
}
```

Doesn't seem to change anything.

**Comment by SchnozzleCat** [comment]
> Hmm I actually have no idea... I have spent an hour or two now trying to look through the source code for neovim, roslyn and c# dev-kit, but I can't see anything that indicates that this shouldn't work... Let me know if you manage to find something

Yeah it's strange. I even compiled the language server from source and set the default for that option to use ``FullSolution``, and it still doesn't work. 

I have no idea how the event handlers for the LSP client work, maybe it's missing one for Workspace Diagnostics or something?

**Comment by seblj** [comment]
Maybe🤔 but I couldn't find anything that would suggest this. It is responding to a `textDocument/diagnostic`, and I assumed that the server returns all diagnostics if the option is on. However, it clearly isn't... I have tried a couple of times to compile it myself to debug, but I havent't managed to do it. How did you do it? I think if we follow everything from when the server gets the request, we could figure out why. However, this is super difficult by just looking at the source code, and without actually debugging it

**Comment by SchnozzleCat** [comment]
> Maybe🤔 but I couldn't find anything that would suggest this. It is responding to a `textDocument/diagnostic`, and I assumed that the server returns all diagnostics if the option is on. However, it clearly isn't... I have tried a couple of times to compile it myself to debug, but I havent't managed to do it. How did you do it? I think if we follow everything from when the server gets the request, we could figure out why. However, this is super difficult by just looking at the source code, and without actually debugging it

I pulled the ``roslyn`` main branch, installed dotnet9, and ran ``dotnet build ./src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Microsoft.CodeAnalysis.LanguageServer.csproj --configuration RELEASE``. 

I assume running it in DEBUG should let you attach a debugger like ``netcoredbg`` to it, but I haven't tested it yet.

**Comment by seblj** [comment]
Okay thanks! Will look at this more at a later time. Feel free the comment here about your findings if you keep debugging this!

**Comment by SchnozzleCat** [comment]
No specific findings, but if you compile with DEBUG you can attach, and if you run the server with the ``--debug`` flag, it will wait for a managed debugger to attach before continuing, which should make debugging easier.

I'll look into it a bit myself, but I don't really know how these language servers work, so unsure if I'll find anything useful in the time I have.

**Comment by seblj** [comment]
I am getting a bunch of these errors trying to compile
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/d8cbe108-8ecc-4c67-8823-63fdc34f8a35">

**Comment by SchnozzleCat** [comment]
That's odd. It looks like it's trying to pull in dependencies from nuget instead of locally, maybe that's why its failing since the code does not match? Not a nuget expert though...

**Comment by seblj** [comment]
Yeah I have no idea. I was trying lots of different things for over an hour now, but no luck... I came across an issue on the repo which looked similar to my issue. However, it was closed as completed and didn't find any workaround or anything

**Comment by franroa** [comment]
In the meanwhile I am using a parserr to get the errors in a quick fix list

**Comment by GustavEikaas** [comment]
I got it working! (kind of)


```lua
return {
  "seblj/roslyn.nvim",
  commit = "5e36cac9371d014c52c4c1068a438bdb7d1c7987",
  config = function()
    require("roslyn").setup({
      config = {
        settings = {
          ["csharp|background_analysis"] = {
            dotnet_compiler_diagnostics_scope = "fullSolution"
          },
          ["csharp|inlay_hints"] = {
            csharp_enable_inlay_hints_for_implicit_object_creation = true,
            csharp_enable_inlay_hints_for_implicit_variable_types = true,
            csharp_enable_inlay_hints_for_lambda_parameter_types = true,
            csharp_enable_inlay_hints_for_types = true,
            dotnet_enable_inlay_hints_for_indexer_parameters = true,
            dotnet_enable_inlay_hints_for_literal_parameters = true,
            dotnet_enable_inlay_hints_for_object_creation_parameters = true,
            dotnet_enable_inlay_hints_for_other_parameters = true,
            dotnet_enable_inlay_hints_for_parameters = true,
            dotnet_suppress_inlay_hints_for_parameters_that_differ_only_by_suffix = true,
            dotnet_suppress_inlay_hints_for_parameters_that_match_argument_name = true,
            dotnet_suppress_inlay_hints_for_parameters_that_match_method_intent = true,
          },
          ["csharp|code_lens"] = {
            dotnet_enable_references_code_lens = true,
          },
        }
      },
      exe = {
        "dotnet",
        vim.fs.joinpath(vim.fn.stdpath("data"), "roslyn", "Microsoft.CodeAnalysis.LanguageServer.dll"),
      },
      filewatching = true,
    })


    vim.keymap.set("n", "<leader>p", function()
      local clients = vim.lsp.get_clients()
      for _, value in ipairs(clients) do
        if value.name == "roslyn" then
          vim.notify("roslyn client found")
          value.rpc.request("workspace/diagnostic", { previousResultIds = {} }, function(err, result)
            if err ~= nil then
              print(vim.inspect(err))
            end
            if result ~= nil then
              print(vim.inspect(result))
            end
          end)
        end
      end
    end, { noremap = true, silent = true })
  end
}
```


1. Use the config above
2. introduce a build error
3. Open any file in the solution
4. `<leader>p`
5. `:messages`
6. voila

Im not sure if vim has some sort of inbuilt lsp support for workspace diagnostics but this should prove that the server itself behaves correctly

**Comment by GustavEikaas** [comment]
@seblj 
Im working on another [dotnet-plugin](https://github.com/GustavEikaas/easy-dotnet.nvim) where I could make good use of the `workspace/symbol` the only downside is that I would require the language server to be running independently of any buffer. If you are interested it would be cool to make a PR that exposes a way to start the lsp server when entering vim if it finds a solution/csproject file.

The benefits of this would be:

1. Project initialization would start when entering vim instead of when entering a file(percieved perf boost)
2. Reduced lag when entering a .cs file because the server is already running
3. Enabling the development of things like workspace diagnostic windows/lualine indicators.

This would of course be exposed as opt-in functionality

**Comment by GustavEikaas** [comment]
@SchnozzleCat @franroa 
Here is a complete qf list keybinding for compiler errors

```lua
return {
  "seblj/roslyn.nvim",
  commit = "5e36cac9371d014c52c4c1068a438bdb7d1c7987",
  config = function()
    require("roslyn").setup({
      config = {
        settings = {
          ["csharp|background_analysis"] = {
            dotnet_compiler_diagnostics_scope = "fullSolution"
          },
          ["csharp|inlay_hints"] = {
            csharp_enable_inlay_hints_for_implicit_object_creation = true,
            csharp_enable_inlay_hints_for_implicit_variable_types = true,
            csharp_enable_inlay_hints_for_lambda_parameter_types = true,
            csharp_enable_inlay_hints_for_types = true,
            dotnet_enable_inlay_hints_for_indexer_parameters = true,
            dotnet_enable_inlay_hints_for_literal_parameters = true,
            dotnet_enable_inlay_hints_for_object_creation_parameters = true,
            dotnet_enable_inlay_hints_for_other_parameters = true,
            dotnet_enable_inlay_hints_for_parameters = true,
            dotnet_suppress_inlay_hints_for_parameters_that_differ_only_by_suffix = true,
            dotnet_suppress_inlay_hints_for_parameters_that_match_argument_name = true,
            dotnet_suppress_inlay_hints_for_parameters_that_match_method_intent = true,
          },
          ["csharp|code_lens"] = {
            dotnet_enable_references_code_lens = true,
          },
        }
      },
      exe = {
        "dotnet",
        vim.fs.joinpath(vim.fn.stdpath("data"), "roslyn", "Microsoft.CodeAnalysis.LanguageServer.dll"),
      },
      filewatching = true,
    })


    vim.keymap.set("n", "<leader>p", function()
      local clients = vim.lsp.get_clients()
      for _, value in ipairs(clients) do
        if value.name == "roslyn" then
          vim.notify("roslyn client found")
          value.rpc.request("workspace/diagnostic", { previousResultIds = {} }, function(err, result)
            if err ~= nil then
              print(vim.inspect(err))
            end
            if result ~= nil then
              local diags = {}
              local seen = {}
              for _, diag in ipairs(result.items) do
                local filepath = diag.uri:gsub("file:///", "")
                if #diag.items > 0 then
                  for _, diag_line in ipairs(diag.items) do
                    if diag_line.severity == 1 then
                      local hash = diag_line.message .. diag_line.range.start.line .. diag_line.range.start.character
                      if seen[hash] == nil then
                        local s = {
                          text = diag_line.message,
                          lnum = diag_line.range.start.line,
                          col = diag_line.range.start.character,
                          filename = filepath
                        }
                        table.insert(diags, s)
                        seen[hash] = true
                      end
                    end
                  end
                end
              end
              vim.fn.setqflist(diags)
              vim.cmd("copen")
            end
          end)
        end
      end
    end, { noremap = true, silent = true })
  end
}

```
![image](https://github.com/user-attachments/assets/f2ad2114-aab8-46fa-ba5d-631ef6900f2f)

**Comment by seblj** [comment]
> @seblj Im working on another [dotnet-plugin](https://github.com/GustavEikaas/easy-dotnet.nvim) where I could make good use of the `workspace/symbol` the only downside is that I would require the language server to be running independently of any buffer. If you are interested it would be cool to make a PR that exposes a way to start the lsp server when entering vim if it finds a solution/csproject file.
> 
> The benefits of this would be:
> 
> 1. Project initialization would start when entering vim instead of when entering a file(percieved perf boost)
> 2. Reduced lag when entering a .cs file because the server is already running
> 3. Enabling the development of things like workspace diagnostic windows/lualine indicators.
> 
> This would of course be exposed as opt-in functionality

Yes I would accept a PR for this! I have actually been thinking about this myself, but haven't decided if I want to use it myself, so I haven't looked into implementing it

I have also been wondering about changing a bit about how it starts, and what it should prefer from a sln or csproj.

As of now, it always finds the sln file and uses that over any csproj, even in subdirectories. As of now it is impossible to load a single project, even if mye cwd is inside that project.

I have been thinking if I should change this to stop searching at cwd or prefer to attach to a project in the cwd rather than a solution in the parent dir.

What do you think?

**Comment by GustavEikaas** [comment]
My personal experience from projects is usually these scenarios

1. There is a sln file and multiple csprojects in subdirectories below it
2. There is a single csproject file

In all the projects I have ever worked on in C# I would always prefer roslyn to pick the broadest scope for my project. So solution is always preferable. The only reason I see for picking a csproject directly instead of a sln file is performance in massive projects. 

`As of now it is impossible to load a single project, even if mye cwd is inside that project.` Given there doesnt exist a solution file this works right?

TLDR; I think your current approach appeals to the broadest audience

**Comment by seblj** [comment]
> If there isnt a solution file then a csproject file is attached right?

Yes

> In all the projects I have ever worked on in C# I would always prefer roslyn to pick the broadest scope for my project. So solution is always preferable. The only reason I see for picking a csproject directly instead of a sln file is performance in massive projects.

Yeah I agree, but I was thinking that if neovim is opened inside one of the projects, then it could use the csproj file, but of course if neovim is opened in the root where the sln file lives, then it would of course use that.

It seems like this could be more consistent with how other editor that has a concept of workspaces do it. Because if I open a specific project in vscode, then it only uses that. But if I open the entire solution, then it used that

**Comment by GustavEikaas** [comment]
Oh now I understand, thats the way my plugin works, never looks up always looks down. Traversing upwards is nice if it doesnt find anything further down I guess. 

Yeah that makes sense to prefer the .csproject if its in cwd or further down and no sln file is present in cwd or inbetween

**Comment by seblj** [comment]
Yes, I might change to this behaviour sometime in the future since I have wished for this a couple of times now. Maybe I will implement something (or ectend) CSTarget to be able to switch to csproj at runtime as well.

Need to think about this a little bit though, and especially the CSTarget thing

**Comment by GustavEikaas** [comment]
Looked into this and made a rough implementation but Im uncertain why its not behaving correctly

1. I start the server
2. I connect the lsp client
3. I send "Initialize" request
4. I send "solution/open"
5. I query diagnostics with a known build error
6. I get 0 results

Doing the same thing with Roslyn starting normally from any buffer in the solution returns the build error.

Ill keep investigating maybe i need to do some diagnostic refresh request or something

**Comment by seblj** [comment]
Okay, so I see now that a handler for `workspace/diagnostic` is just not implement in neovim core. I tried to search for an issue there or something to see if I could find some discussions about it or something, but I couldn't find it.

I am unfortunately not going to implement the handler in this plugin, as I believe efforts to support the handler should instead go into neovim core. I don't think I will look into sending a PR over there myself.

So with that said, I don't think there is anything to do for me right now for this, but feel free to paste snippets here if there are some improvements to the snippets already here or something, so that other users could take advantage of them😄

**Comment by seblj** [comment]
BTW: it is not workspace diagnostics, but it is a pretty nice builtin feature to vim that is not so well known.

You can set `:help compiler` to `dotnet` with `:compiler dotnet`, and then you should be able to run `:make`. Then all diagnostics that is outputted from a `dotnet build` is shown in the qflist.

Just thought I could share it here as a "workaround"

**Comment by cphorton** [comment]
Go and give this issue some love guys https://github.com/neovim/neovim/issues/34231

**Comment by Eckii24** [comment]
https://github.com/neovim/neovim/pull/34262 was just merged today and resolves https://github.com/neovim/neovim/issues/34231

Not sure, if this will help here. Would love to see workspace diagnostics work.

**Comment by froats** [comment]
Hi @ZieMcd  did you got something? I also not able to get it working

**Comment by ZieMcd** [comment]
Hi, from what I can see it looks like neovim sends an ResultId with a null DocumentUri when making the workspace diagnostic request which causes the lsp to throw an `Object reference not set to an instance of an object` error. 
I think the error is in neovim since the spec does not indicate that ResultId.DocumentUri can be null. Will raise an issue in neovim or maybe even try fix it myself

**Comment by ZieMcd** [comment]
The PR has been merged and vim.lsp.buf.workspace_diagnostics() does now work. Only issue is that the lsp does not send workspace diagnostics as part of it's capabilities. I'm not sure why, I assume there is some setting we have to set ... maybe.

**Comment by PaddiM8** [comment]
Huh, strange. Can't see it being enabled here: https://github.com/dotnet/roslyn/blob/ad0a3e5531230c6f934c1cad8ff9365f929476db/src/LanguageServer/Protocol/DefaultCapabilitiesProvider.cs

And the language *client* seems to enable it itself in its own instance of the server capabilities object..?

https://github.com/dotnet/roslyn/blob/ad0a3e5531230c6f934c1cad8ff9365f929476db/src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs#L74

Maybe we'll have to patch the server capabilities response client side but I don't know

**Comment by seblyng** [comment]
Hmm, I am a bit unsure about that. It feels like there must be a reason for them to not specify that it is a capability that they support. Is this something that is working in vscode? If so, how?

It's at least very easy to hack it with `vim.lsp.config`:

```lua
vim.lsp.config("roslyn", {
    on_attach = function(client)
        client.server_capabilities.diagnosticProvider = { workspaceDiagnostics = true }
    end,
})
```

**Comment by franroa** [comment]
@ZieMcd can you share your config? That hack is not working for me

**Comment by seblyng** [comment]
Hmm yeah I didn't really test it when I suggested using `on_attach`, and I also can't get it to work

**Comment by franroa** [comment]
Not working for me, with this config:
```

vim.lsp: Active Clients ~
- roslyn (id: 1)
  - Version: ? (no serverInfo.version response)
  - Root directory: ~/Projects/learning/remapp-5.0
  - Command: { "/home/froa/.local/share/nvim/mason/bin/roslyn", "--logLevel=Information", "--extensionLogDirectory=/home/froa/.local/state/nvim", "--stdio" }
  - Settings: {
      ["csharp|background_analysis"] = {
        dotnet_analyzer_diagnostics_scope = "fullSolution",
        dotnet_compiler_diagnostics_scope = "fullSolution"
      },
      ["csharp|code_lens"] = {
        dotnet_enable_references_code_lens = true,
        dotnet_enable_tests_code_lens = true
      },
      ["csharp|completion"] = {
        dotnet_provide_regex_completions = true,
        dotnet_show_completion_items_from_unimported_namespaces = true,
        dotnet_show_name_completion_suggestions = true
      },
      ["csharp|highlighting"] = {
        dotnet_highlight_related_json_components = true,
        dotnet_highlight_related_regex_components = true
      },
      ["csharp|implement_type"] = {
        dotnet_insertion_behavior = true,
        dotnet_property_generation_behavior = true
      },
      ["csharp|inlay_hints"] = {
        csharp_enable_inlay_hints_for_implicit_object_creation = true,
        csharp_enable_inlay_hints_for_implicit_variable_types = true,
        csharp_enable_inlay_hints_for_lambda_parameter_types = true,
        csharp_enable_inlay_hints_for_types = true,
        dotnet_enable_inlay_hints_for_indexer_parameters = true,
        dotnet_enable_inlay_hints_for_literal_parameters = true,
        dotnet_enable_inlay_hints_for_object_creation_parameters = true,
        dotnet_enable_inlay_hints_for_other_parameters = true,
        dotnet_enable_inlay_hints_for_parameters = true,
        dotnet_suppress_inlay_hints_for_parameters_that_differ_only_by_suffix = true,
        dotnet_suppress_inlay_hints_for_parameters_that_match_argument_name = true,
        dotnet_suppress_inlay_hints_for_parameters_that_match_method_intent = true
      },
      ["csharp|navigation"] = {
        dotnet_navigate_to_decompiled_sources = true
      },
      ["csharp|quick_info"] = {
        dotnet_show_remarks_in_quick_info = true
      },
      ["csharp|symbol_search"] = {
        dotnet_search_reference_assemblies = true
      }
    }
  - Attached buffers: 10

```

**Comment by ZieMcd** [comment]
hmm, are you running the latest nightly build of nvim?
trying running this command `:lua print(vim.inspect(vim.lsp.get_client_by_id(your_roslyn_server_id).server_capabilities))`
is this included in result ?
```
  diagnosticProvider = {                                                                                                                                                                                                                                     
    workspaceDiagnostics = true                                                                                                                                                                                                                              
  },  
```

**Comment by PaddiM8** [comment]
I asked in the C# discord server and actually got a response from someone that has worked with this. Apparently it only supports dynamic registration, which seems to be disabled for diagnostics in neovim:

https://github.com/neovim/neovim/blob/master/runtime%2Flua%2Fvim%2Flsp%2Fprotocol.lua

Don't have the possibility to try it out right now but maybe enabling that would do the trick? Not sure if you have to do something else as well or if there is sufficient support in the neovim lsp client though 

They seem to be having the same problem in sublime
 https://github.com/sublimelsp/LSP/issues/2561

**Comment by ZieMcd** [comment]
looks like we do as a hack
https://github.com/seblyng/roslyn.nvim/blob/3056635ed6ebe5035c94a7c8d0d362e95817839e/lsp/roslyn.lua#L35
but I guess Neovim doesn't implement it properly.

**Comment by franroa** [comment]
I do have that in my print:
```

12:58:15 msg_show.lua_print   :lua print(vim.inspect(vim.lsp.get_client_by_id(1).server_capabilities)) {
  _vs_onAutoInsertProvider = {
    _vs_triggerCharacters = { "'", "/", "\n" }
  },
  codeActionProvider = {
    codeActionKinds = { "quickfix", "refactor" },
    resolveProvider = true
  },
  codeLensProvider = {
    resolveProvider = true
  },
  completionProvider = {
    allCommitCharacters = { " ", "{", "}", "[", "]", "(", ")", ".", ",", ":", ";", "+", "-", "*", "/", "%", "&", "|", "^", "!", "~", "=", "<", ">", "?", "@", "#", "'", '"', "\\" },
    resolveProvider = true,
    triggerCharacters = { '"', "(", ":", "[", "\\", "{", "#", ".", ">", " ", "~", "<" }
  },
  definitionProvider = true,
  diagnosticProvider = {
    workspaceDiagnostics = true
  },
```
